coderd/workspaceapps.go

```diff
@@ -31,60 +33,144 @@ func (api *API) workspaceAppsProxyPath(rw http.ResponseWriter, r *http.Request)
+	// rendered on error. This should be set for proxy path URLs but not
+	// hostname based URLs.
+	DashboardOnError bool
+}
+
+func (api *API) proxyWorkspaceApplication(proxyApp proxyApplication, rw http.ResponseWriter, r *http.Request) {
+	if !api.Authorize(r, rbac.ActionCreate, proxyApp.Workspace.ExecutionRBAC()) {
+		httpapi.ResourceNotFound(rw)
 		return
 	}
-	if !app.Url.Valid {
-		httpapi.Write(rw, http.StatusBadRequest, codersdk.Response{
-			Message: fmt.Sprintf("Application %s does not have a url.", app.Name),
+
+	// If the app does not exist, but the app name is a port number, then
+	// route to the port as an "anonymous app". We only support HTTP for
+	// port-based URLs.
+	internalURL := fmt.Sprintf("http://127.0.0.1:%d", proxyApp.Port)
+
+	// If the app name was used instead, fetch the app from the database so we
+	// can get the internal URL.
+	if proxyApp.AppName != "" {
+		app, err := api.Database.GetWorkspaceAppByAgentIDAndName(r.Context(), database.GetWorkspaceAppByAgentIDAndNameParams{
+			AgentID: proxyApp.Agent.ID,
+			Name:    proxyApp.AppName,
 		})
-		return
+		if err != nil {
+			httpapi.Write(rw, http.StatusInternalServerError, codersdk.Response{
+				Message: "Internal error fetching workspace application.",
+				Detail:  err.Error(),
+			})
+			return
+		}
+
+		if !app.Url.Valid {
+			httpapi.Write(rw, http.StatusBadRequest, codersdk.Response{
+				Message: fmt.Sprintf("Application %s does not have a url.", app.Name),
+			})
+			return
+		}
+		internalURL = app.Url.String
 	}
 
-	appURL, err := url.Parse(app.Url.String)
+	appURL, err := url.Parse(internalURL)
 	if err != nil {
 		httpapi.Write(rw, http.StatusInternalServerError, codersdk.Response{
-			Message: fmt.Sprintf("App url %q must be a valid url.", app.Url.String),
+			Message: fmt.Sprintf("App URL %q is invalid.", internalURL),
 			Detail:  err.Error(),
 		})
 		return
 	}
 
-	proxy := httputil.NewSingleHostReverseProxy(appURL)
-	proxy.ErrorHandler = func(w http.ResponseWriter, r *http.Request, err error) {
-		// This is a browser-facing route so JSON responses are not viable here.
-		// To pass friendly errors to the frontend, special meta tags are overridden
-		// in the index.html with the content passed here.
-		r = r.WithContext(site.WithAPIResponse(r.Context(), site.APIResponse{
-			StatusCode: http.StatusBadGateway,
-			Message:    err.Error(),
-		}))
-		api.siteHandler.ServeHTTP(w, r)
-	}
-	path := chi.URLParam(r, "*")
-	if !strings.HasSuffix(r.URL.Path, "/") && path == "" {
+	// Ensure path and query parameter correctness.
+	if proxyApp.Path == "" {
 		// Web applications typically request paths relative to the
 		// root URL. This allows for routing behind a proxy or subpath.
 		// See https://github.com/coder/code-server/issues/241 for examples.
-		r.URL.Path += "/"
-		http.Redirect(rw, r, r.URL.String(), http.StatusTemporaryRedirect)
+		http.Redirect(rw, r, r.URL.Path+"/", http.StatusTemporaryRedirect)
```

Comment:
`http.Redirect` expects a URL but you've supplied a path here.

Reply:
I don't think `http.Redirect` expects a full URL and works fine with relative or absolute paths. From godoc:
> Redirect replies to the request with a redirect to url, which may be a path relative to the request path.

Reply:
Ah that's nice, wasn't sure

Reply:
I just tried this program to double check and it seems to work.

```go
package main

import (
	"fmt"
	"net/http"
)

func main() {
	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		http.Redirect(w, r, "/hi", http.StatusTemporaryRedirect)
	})
	http.HandleFunc("/hi", func(w http.ResponseWriter, r *http.Request) {
		fmt.Fprintf(w, "hi")
	})

	panic(http.ListenAndServe(":8080", nil))
}
```